### ISSUE TITLE
Installation issue: py-h5py@master+mpi ^hdf5@develop-1.15

### ISSUE THREAD

**Issue by hyoklee** [LLNL/spack#37864]
### Steps to reproduce the issue

```console
$ spack spec -I py-h5py@master+mpi ^hdf5@develop-1.15
...
```

### Error message

<details><summary>Error message</summary>

<pre>
==> Error: concretization failed for the following reasons:

   1. Cannot select a single "version" for package "hdf5"
   2. Cannot satisfy 'hdf5@develop-1.15'
   3. Cannot satisfy 'hdf5@1.8.4:1.14'
   4. Cannot satisfy 'hdf5@1.6:1.14'

</pre></details>

### Information on your system

* **Spack:** 0.21.0.dev0 (91a54029f9980a6b08c66263d0b69b6980cf1844)
* **Python:** 3.11.3
* **Platform:** darwin-monterey-skylake
* **Concretizer:** clingo

### Additional information

@adamjstewart  @bryanherman  @pradyunsg  @takluyver

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by takluyver** [comment]
The Spack h5py package specifies compatibility with HDF5 up to 1.14:

```python
depends_on("hdf5@1.8.4:1.14 +hl", when="@3.8:")
```

We expect that some changes will be necessary to support 1.16 when that comes out, and h5py doesn't really aim to support the odd-numbered development releases of HDF5. They might work, depending on what changes in HDF5, but it's not really a priority for us to ensure that they do.

**Comment by hyoklee** [comment]
Yes, hdf5 develop works on h5py master. I wish you could allow testing h5py against hdf5 develop via Spack.

**Comment by takluyver** [comment]
From my perspective, you're allowed to test that, but I'd prefer the package didn't advertise compatibility with new versions of HDF5 until it's known to work. I'm not familiar enough with Spack to know if it has some option to override a particular constraint.

**Comment by adamjstewart** [comment]
There is no option to override `depends_on`, and unfortunately there's also no `suggests` directive with soft constraints. If you want to experiment and you don't mind hacking, you can always `spack edit py-h5py` and comment out the upper bound. But I agree with @takluyver that stability is important for these kinds of packages.

**Comment by hyoklee** [comment]
Thanks, @adamjstewart , for the tip.

In this case,  should I try this?

```
depends_on("hdf5@1.8.4: +hl", when="@3.8:")
```

**Comment by adamjstewart** [comment]
Yes, that should work for experimentation
